File: leetcode/letter-tile-possibilities.java
class Solution {
    int count = 0;

    public void traverse(char theArray[], boolean used[]) {
        count++;
        for(int i = 0; i < theArray.length; i++) {
            if( used[i] || (i > 0 && theArray[i] == theArray[i - 1] && !used[i - 1]) )
                continue;
            used[i] = true;
            traverse(theArray, used);
            used[i] = false;
        }
    }

    public int numTilePossibilities(String tiles) {
        char theArray[] = tiles.toCharArray(); 
        Arrays.sort(theArray);
        traverse(theArray, new boolean[theArray.length]);
        return count - 1;
    }
}

File: leetcode-questions/letter-tile-possibilities.cpp
class Solution {
public:
    int traverse(string tiles, map<int,int> cc) {
        int sum = 0;
        for (int i = 0; i < 26; i++) {
            if(cc[i]==0) continue;
            sum++;
            cc[i]--;
            sum += traverse(tiles, cc);
            cc[i]++;
        }
        return sum;
    }
    int numTilePossibilities(string tiles) {
        map<int,int> cc;
        sort(tiles.begin(), tiles.end());
        for(char c: tiles) {
            cc[c-'A'] += 1;
        }
        return traverse(tiles, cc);
    }
};